cfgmgr/buffermgr.cpp

```diff
@@ -0,0 +1,201 @@
+    {
+        SWSS_LOG_ERROR("Unable to create/update PG profile for port %s. No PG profile configured for speed %s and cable length %s",
+                       port.c_str(), speed.c_str(), cable.c_str());
+        return;
+    }
+
+    // Crete record in BUFFER_PROFILE table
+    // key format is pg_lossless_<speed>_<cable>_profile
+    string buffer_pg_key = port + CONFIGDB_TABLE_NAME_SEPARATOR + LOSSLESS_PGS;
+    string buffer_profile_key = "pg_lossless_" + speed + "_" + cable + "_profile";
+
+    // check if profile already exists - if yes - skip creation
+    m_cfgBufferProfileTable.get(buffer_profile_key, fvVector);
+    if (fvVector.size() == 0)
+    {
+        SWSS_LOG_NOTICE("Creating new profile '%s'", buffer_profile_key.c_str());
+
+        string mode = getPgPoolMode();
+        if (mode.empty())
+        {
+            // this should never happen if switch initialized properly
+            SWSS_LOG_ERROR("PG lossless pool is not yet created");
+            return;
+        }
+
+        // profile threshold field name
+        mode += "_th";
+        string pg_pool_reference = string(CFG_BUFFER_POOL_TABLE_NAME) + CONFIGDB_TABLE_NAME_SEPARATOR + INGRESS_LOSSLESS_PG_POOL_NAME;
+        fvVector.push_back(make_pair("pool", "[" + pg_pool_reference + "]"));
+        fvVector.push_back(make_pair("xon", m_pgProfileLookup[speed][cable].xon));
+        fvVector.push_back(make_pair("xoff", m_pgProfileLookup[speed][cable].xoff));
+        fvVector.push_back(make_pair("size", m_pgProfileLookup[speed][cable].size));
+        fvVector.push_back(make_pair(mode, m_pgProfileLookup[speed][cable].threshold));
+        m_cfgBufferProfileTable.set(buffer_profile_key, fvVector);
+    }
+    else
+    {
+        SWSS_LOG_NOTICE("Reusing existing profile '%s'", buffer_profile_key.c_str());
+    }
+
+    fvVector.clear();
+    string profile_ref = string("[") + CFG_BUFFER_PROFILE_TABLE_NAME + CONFIGDB_TABLE_NAME_SEPARATOR + buffer_profile_key + "]";
+    fvVector.push_back(make_pair("profile", profile_ref));
+    m_cfgBufferPgTable.set(buffer_pg_key, fvVector);
+}
+
+void BufferMgr::doTask(Consumer &consumer)
+{
+    SWSS_LOG_ENTER();
+
+    string table_name = consumer.getTableName();
+
+    auto it = consumer.m_toSync.begin();
+    while (it != consumer.m_toSync.end())
+    {
+        KeyOpFieldsValuesTuple t = it->second;
+
+        string keySeparator = CONFIGDB_KEY_SEPARATOR;
+        vector<string> keys = tokenize(kfvKey(t), keySeparator[0]);
+        string port(keys[0]);
+
+        string op = kfvOp(t);
+        if (op == SET_COMMAND)
+        {
+            for (auto i : kfvFieldsValues(t))
+            {
+                if (table_name == CFG_PORT_CABLE_LEN_TABLE_NAME)
+                {
+                    // receive and cache cable length table
+                    doCableTask(fvField(i), fvValue(i));
+                }
+                if (table_name == CFG_PORT_TABLE_NAME && fvField(i) == "speed")
+                {
```

Comment:
Will the field ever != "speed"?

If yes, why ignoring them is correct? Please at least add comment to state reason why it is safe to ignore them.

Reply:
yes, we receive updates for other fields as well. Actually every time at least one field in the table gets updated, all subscribers receive notifications about all fields. buffer manager is interested in speed only. I'll add a comment.

Reply:
Thanks for the explanation!